src/vmm/src/device_manager/mod.rs

```diff
@@ -15,3 +47,422 @@ pub mod mmio;
+        };
+
+        Ok(DeviceManager {
+            resource_allocator: ResourceAllocator::new()?,
+            mmio_bus,
+            mmio_devices: MMIODeviceManager::new(),
+            #[cfg(target_arch = "x86_64")]
+            pio_bus,
+            #[cfg(target_arch = "x86_64")]
+            legacy_devices,
+            acpi_devices: ACPIDeviceManager::new(),
+        })
+    }
+
+    /// Attaches a VirtioDevice device to the device manager and event manager.
+    pub(crate) fn attach_virtio_device<T: 'static + VirtioDevice + MutEventSubscriber + Debug>(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        vmfd: &VmFd,
+        id: String,
+        device: Arc<Mutex<T>>,
+        cmdline: &mut Cmdline,
+        is_vhost_user: bool,
+    ) -> Result<(), AttachMmioDeviceError> {
+        let interrupt = Arc::new(IrqTrigger::new());
+        // The device mutex mustn't be locked here otherwise it will deadlock.
+        let device = MmioTransport::new(mem.clone(), interrupt, device, is_vhost_user);
+        self.mmio_devices.register_mmio_virtio_for_boot(
+            vmfd,
+            &mut self.resource_allocator,
+            id,
+            &self.mmio_bus,
+            device,
+            cmdline,
+        )?;
+
+        Ok(())
+    }
+
+    /// Attaches a [`BootTimer`] to the VM
+    pub(crate) fn attach_boot_timer_device(
+        &mut self,
+        request_ts: TimestampUs,
+    ) -> Result<(), AttachMmioDeviceError> {
+        let boot_timer = Arc::new(Mutex::new(BootTimer::new(request_ts)));
+
+        self.mmio_devices.register_mmio_boot_timer(
+            &self.mmio_bus,
+            &mut self.resource_allocator,
+            boot_timer,
+        )?;
+
+        Ok(())
+    }
+
+    pub(crate) fn attach_vmgenid_device(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        vmfd: &VmFd,
+    ) -> Result<(), AttachVmgenidError> {
+        let vmgenid = VmGenId::new(mem, &mut self.resource_allocator)?;
+        self.acpi_devices.attach_vmgenid(vmgenid, vmfd)?;
+        Ok(())
+    }
+
+    #[cfg(target_arch = "aarch64")]
+    pub(crate) fn attach_legacy_devices_aarch64(
+        &mut self,
+        vmfd: &VmFd,
+        event_manager: &mut EventManager,
+        cmdline: &mut Cmdline,
+    ) -> Result<(), AttachLegacyMmioDeviceError> {
+        // Serial device setup.
+        let cmdline_contains_console = cmdline
+            .as_cstring()
+            .map_err(|_| AttachLegacyMmioDeviceError::Cmdline)?
+            .into_string()
+            .map_err(|_| AttachLegacyMmioDeviceError::Cmdline)?
+            .contains("console=");
+
+        if cmdline_contains_console {
+            // Make stdout non-blocking.
+            Self::set_stdout_nonblocking();
+            let serial = Self::setup_serial_device(event_manager)?;
+            self.mmio_devices.register_mmio_serial(
+                vmfd,
+                &self.mmio_bus,
+                &mut self.resource_allocator,
+                serial.clone(),
```

Comment:
Unneeded clone. And same for rtc bellow.

Reply:
fixed